packages/core/src/StateNode.ts

```diff
@@ -707,14 +708,25 @@ class StateNode<
    * @param state The state to resolve
    */
   public resolveState(
-    state: State<TContext, TEvent, any, any>
+    state: State<TContext, TEvent, any, any> | StateConfig<TContext, TEvent>
   ): State<TContext, TEvent, TStateSchema, TTypestate> {
+    const resolvedState =
+      state instanceof State
+        ? state
+        : isStateConfig(state)
+        ? State.create(state)
+        : undefined;
+
+    if (!resolvedState) {
+      throw new Error(`Cannot resolve state: invalid state object`);
+    }
```

Comment:
Those lines should kinda not be needed at all - the argument is typed and thus an assignment like this would already be exhaustive:
```js
const resolvedArg = state instanceof State ? state : State.create(state)
```

Reply:
They are needed because even though the types are exhaustive, data coming in here is often from an untrustworthy source, and needs to be checked in runtime:

```js
const stateConfig = JSON.parse(localStorage.get('state')); // cannot guarantee that this is a valid state config!

machine.resolveState(stateConfig); // might fail, despite types not failing
```

Reply:
The counter-point is that... this can probably still fail or result in an invalid/incomplete value despite those checks, so this is not buying us that much here.

Reply:
> The counter-point is that... this can probably still fail or result in an invalid/incomplete value despite those checks, so this is not buying us that much here.

Okay, removed